USB: Implement USB Ethernet Device gadget (ECM protocol)

finikorg opened the pull request:
First revision of CDC ECM gadget device.

TODO: Re-use composite device headers and implement common register_function() methods

Commit:
Revert "cdc_acm : Restrict writing more than 4 bytes into TX USB Endpoint."

This reverts commit 1da0a9eebd8b1a15e8247e0fb360171599d981e5.

The workaround caused a severe performance penalty, and only worked
for USB packets of 4-15 bytes in length (16+ byte packets weren't
subject to the hardware bug).  Single-byte packets (very common for
cdc_acm serial port transfers) would still be duplicated sometimes.

The upcoming DMA implementation does not share the performance
penalty, and also is not subject to the bug for those sizes of packets
(though it DOES still have a problem with single-byte packets!).

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
usb_dc_dw: Correctly assign endpoint FIFOs

The designware hardware in dedicated FIFO mode (which is all we
support right now for lack of shared-FIFO hardware) has one hardware
FIFO per IN (i.e. transmit) endpoint.  But it doesn't assign them on
its own, it's the drivers responsibility to populate the TxFNum field
of the DIEPCTL registers with integer indices corresponding to the
desired FIFO.

We weren't doing that, which meant that all IN endpoints were sharing
the same FIFO zero which is supposed to be dedicated to EP0 control
transfers.  The net effect is that sometimes outbound transfers would
be corrupted, showing data from the wrong endpoint.  More often that
not this would leak from control transfers over to the
higher-bandwidth bulk endpoints of the application, but occasionally
you'd see a control transfer itself get borked and the USB device
would glitch.

Get this right and set the FIFOs properly.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
usb_dc_dw: Correctness: use bytewise I/O in FIFO fill

The Designware FIFO is filled in units of 32 bit words, but the buffer
we are passed is not guaranteed to be a multiple of 4 bytes long, nor
aligned on a 4-byte boundary.  So in theory we are reading 0-3 bytes
of unused garbage from the end of the array.

That's currently benign on supported platforms with this hardware,
which all support misaligned reads.  But not all do.  And the incoming
arrival of memory protection opens the possibility that those extra
bytes would cross a protection boundary and cause a crash or security
bug.

Do this right.

(Note that this is fixed to little endian byte order.  The Designware
databook is frustratingly silent on the endianness it expects, but
existing hardware I can see is definitely LE and I see a few spots in
the Linux dwc2 driver that likewise assume LE).

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
usb_dw_dc: Fix locking correctness

There is no particular reason this spot in usb_dw_tx() cannot be
reached by racing threads on the same endpoint, though existing API
usage in the tree is all unithreaded.  The FIFO state read at the top
of the function must still be true at the bottom or else the packet
byte count will be corrupt.

Also, as described in an existing comment, the databook has some
scary-sounding warnings about access to the registers during FIFO
operations, even if they "should" be on separate endpoints and
unrelated.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
usb_dw_dc: Cleanup: write DIEPCTL register just once

The enable and clear NAK bit can be legally written together, do it in
just one write.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

finikorg committed:
usb: Add extra debug information

Add extra debug information to USB commands.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: netusb: Add composite device with ECM function

Add composite device skeleton with ECM function implemented.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: netusb: Add ethernet emulation device

Ethernet emulation device allows to use networking interface for
interaction with USB endpoints.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
net: sample: Add USB Ethernet stick configuration

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg commented:
@jfischer-phytec-iot @pfalcon Please review first revision, I am going to re-use descriptors like in 9c916522b45fefd5891f1ac1fb5d0cb1594067af.

pfalcon commented:
@finikorg : Thanks for the patch! I'm not sure how much of it I'll be able to review though, I mostly keep an eye on USB things because I'd like to have an USB CDC driver for ARM devices to be implemented, but I don't have enough experience doing that myself.

finikorg committed:
usb: Add CDC ECM composite configuration

Add USB device configuration for CDC ECM

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

jfischer-phytec-iot commented on subsys/usb/class/netusb/eth_emu.c:
```diff
@@ -0,0 +1,81 @@
+
+static struct eth_emu_context eth_emu_ctx;
+
+static int eth_emu_init(struct device *dev)
```
eth_emu does not need to depend on usb, I think it would be better to separate it from this PR  and move to drivers/ethernet/, @jukkar your opinion?

finikorg replied:
I can move it to some other place if needed.

finikorg replied:
@jukkar @tbursztyka @rveerama1 could you check this virtual ethernet device?

jfischer-phytec-iot replied:
see also https://github.com/zephyrproject-rtos/zephyr/pull/4461#issuecomment-338216079

jfischer-no reviewed (dismissed):
@finikorg it all looks like WIP,
can you please open an issue and describe your work, like: - add ethernet emulation device, add CDC-ECM driver, add RNDIS driver. I guess ethernet emulator can also be used with CDC-EEM driver and it would be good to synchronize the work on it with @loicpoulain

finikorg commented:
@nashif Actually I am trying to rebase it now and reuse composite device framework.

finikorg commented:
@nashif @jfischer-phytec-iot rebased to work with new interface

finikorg committed:
usb: Add sanity_check configuration for netusb

Add netusb and composite netusb build check configurations.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg commented:
@laperie Please review

jfischer-phytec-iot commented on subsys/usb/usb_descriptor.c:
```diff
@@ -59,6 +59,23 @@ struct dev_common_descriptor {
 #endif
+#ifdef CONFIG_USB_DEVICE_NETWORK_ECM
+	struct usb_cdc_ecm_config {
+		struct usb_association_descriptor iad;
```
association descriptor is necessary for composite device only, see above

finikorg replied:
@jfischer-phytec-iot fixed

jfischer-no reviewed (commented):
I looked at it quickly, will do more closely later.

nashif reviewed (changes requested):
This commit is not increasing the debug level, it is adding more information, so please fix the commit message

nashif commented:
> This commit is not increasing the debug level, it is adding more information, so please fix the commit message

referring to 
```

Author:     Andrei Emeltchenko <andrei.emeltchenko@intel.com>
AuthorDate: Wed Aug 16 16:22:59 2017 +0300
Commit:     Andrei Emeltchenko <andrei.emeltchenko@intel.com>
CommitDate: Fri Oct 20 11:18:17 2017 +0300

    usb: Increase debug level

    Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

```

nashif commented:
then we have a few commits where you 
- first add Kconfig
- then add sources
- then incorporate those into Makefiles

Please squash them, when a source file is introduced, it had to be buildable, no need to split the commits this way.

I am referring to 
```
2017-09-26 15:40 Andrei Emeltchenko             o usb: netusb: Add build infrastructure for USB net device
2017-10-02 13:55 Andrei Emeltchenko             o usb: netusb: Add composite device with ECM function
2017-09-26 15:30 Andrei Emeltchenko             o usb: netusb: Add ethernet emulation device
2017-09-25 13:44 Andrei Emeltchenko             o usb: netusb: Add ECM protocol handling
2017-10-02 13:40 Andrei Emeltchenko             o usb: netusb: Add RNDIS Networking device configuration
2017-10-02 13:36 Andrei Emeltchenko             o usb: netusb: Add ECM Networking Device configuration

```

nashif commented on subsys/usb/class/netusb/Kconfig:
```diff
@@ -0,0 +1,77 @@
+menu "USB Device Networking support"
+	depends on USB_DEVICE_STACK
+
+config USB_DEVICE_NETWORK
```
what is the idea behind this kconfig option? it has no description and is automatically selected

finikorg replied:
@nashif idea here is that we implement some other functions (RNDIS, etc). Common code is built whenever network function is enabled. I have working solution with several functions where it does make sense

nashif commented:
You just added netusb.c, so why are you refactoring it in a followup commit? Why not squash this

`usb: netusb: Refactor netusb code using new interface` 

into where the file was introduced?

When reviewing this commit by commit, it just makes reviews very difficult, please submit less commits that are self-contained rather than bit and pieces in many more commits without  a good reason.

finikorg commented:
@nashif for some reason I cannot reply directly to comments due to this **** github interface, so I reply in my comment here. I squashed most of the commits together. I left emulation device commit separate since it may be reworked to be included to networking codebase.

finikorg commented:
I squashed even more, only Ethernet emulation device left separate.

finikorg commented:
rebased against the latest upstream

finikorg committed:
usb: Use RFC 7042 Documentation values for MAC strings

Use documentation defined values for virtual devices MAC addresses in
Zephyr and Host OS.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: dw: Fix debug log specifier for FIFO not empty

The situation when FIFO is not empty is not a bug and it is spamming
console when only bugs are enabled.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: dw: Add sleep before FIFO flash

Adding sleep before TX FIFO flash fixes splitting networking packets
sent over USB endpoints making ECM broken since there is no flow
control other then frame sizes.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

jfischer-phytec-iot commented:
@finikorg  Andrei, can you please move dw driver relevant commits in a separate PR. I fear that dw driver fixes here will be overlooked and someone begins to fix dw driver, thx

nashif commented:
recheck

lpereira commented on subsys/usb/class/netusb/function_ecm.c:
```diff
@@ -0,0 +1,297 @@
+
+	net_hexdump_frags("<", pkt);
+
+	if (!pkt->frags) {
```
If `pkt->frags` is NULL, the `net_pkt_ll()` call below will crash: it calls `net_pkt_ip_data()`, which dereferences `pkt->frags`.  I could fix this, but I don't know the intent of this code here.

finikorg replied:
This is fixed in PR #4641

jli157 commented:
@finikorg If using a device with this driver enabled on Windows, what kind of windows driver can be used to support the device?

finikorg commented:
> @finikorg If using a device with this driver enabled on Windows, what kind of windows driver can be used to support the device?

I could not make it working in Windows, it is better to use MS RNDIS.
